### PR TITLE
feat: documents state transition rate limiter

### PR DESCRIPTION
**Pull request by Cofresi**
Adding a temporary rate limiter as spam protection. This will probably be removed once the credit system is live.

- The rate limiter is checking only document state transtions on a per identity per time interval basis.
- `Events` (formerly `tags`) are used to check how many document state transition per time interval an identity has made.
- The rate limiter can be activated and its options changed in the .env file

**Note:** 

We need to check if the abci interface won't break in upcoming versions of Tendermint. I think it will since new `events` are replacing `tags`. They already renamed the structure in the newest version and it might still be backwards compatible. But the `events` structure also supports a new `type` property, which seems to be the first part separated by a dot in the search string (e.g. in `ratelimiter.interval.10` `ratelimiter` is a `type`). Probably the `type` field will need to be passed separately to the `event` constructor in a new version of abci.

Cf. https://github.com/tendermint/tendermint/pull/4046

### CODE REVIEW

File: lib/api/TendermintRPCClient.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+   * Fetches transitions by Event (formerly Tag).
+   * A tag is either just a key or a key value combination.
+   * @param {string} key
+   * @param {string} [value='']
+   * @param {boolean} [prove=false]
+   * @param {string} [page='1']
+   * @param {string} [perPage='30']
+   * @return {Promise<Object[]|null>}
+   */
+  async fetchTransitionsByEvent(key, value = '', prove = false, page = '1', perPage = '30') {
```
Comment (shumkov) [inline lib/api/TendermintRPCClient.js]: it's better to have a more business logic oriented method that responds with total_count

File: lib/tendermint/fetchTransitionCountByEventFactory.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+function fetchTransitionCountByEventFactory(tendermintRPC) {
+  /**
+   * Fetch total count of transitions by Event (formerly Tag)
+   * A tag is either just a key or a key-value combination
+   *
+   * @typedef fetchTransitionCountByEvent
+   *
+   * @param {string} key
+   * @param {string} [value='']
+   * @param {boolean} [options.prove=false]
```
Comment (shumkov) [inline lib/tendermint/fetchTransitionCountByEventFactory.js]: These options won't affect query so I would remove them at all.

File: lib/abci/handlers/checkTxHandlerFactory.js [hunk 3]
```diff
@@ -2,42 +2,95 @@ const {
...
       }
-
       throw e;
     }
 
+    // apply rate limiter for document state transitions
+    if (rateLimiterActive && stateTransition.getType() === stateTransitionTypes.DOCUMENTS) {
+      const { userId } = stateTransition.documents[0];
+
+      const isBannedUser = await rateLimiter.isBannedUser(
```
Comment (shumkov) [inline lib/abci/handlers/checkTxHandlerFactory.js]: We should check it for `deliverTx` as well. ST can come with a block, so `checkTx` won't be called.

File: lib/services/RateLimiter.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  /**
+   * checks if transition quota per identity is exceeded
+   * @param {string} userId
+   * @param {number} blockHeight
+   * @return {Promise<boolean>}
+   */
+  async isQuotaExceeded(userId, blockHeight) {
+    const limitedKey = this.getRateLimitedKey(blockHeight);
+    const count = await this.fetchTransitionCountByEvent(limitedKey, userId, { prove: true });
```
Comment (shumkov) [inline lib/services/RateLimiter.js]: Why do we need proves from the local Tendermint node?

File: lib/abci/handlers/checkTxHandlerFactory.js [hunk 5]
```diff
@@ -2,42 +2,95 @@ const {
...
+        userId, blockchainState.getLastBlockHeight(),
+      );
+
+      if (isBannedUser) {
+        const bannedResponse = new ResponseCheckTx();
+
+        bannedResponse.code = AbciError.CODES.RATE_LIMITER_BANNED;
```
Comment (shumkov) [inline lib/abci/handlers/checkTxHandlerFactory.js]: Let's create a new error type `UserIsBannedAbciError` https://github.com/dashevo/js-machine/blob/v0.3-dev/lib/abci/errors/

```
message = `User ${userId} is banned for some time`
data = { userId } 
```

File: lib/abci/handlers/checkTxHandlerFactory.js [hunk 6]
```diff
@@ -2,42 +2,95 @@ const {
...
+      }
+
+      const isQuotaExceeded = await rateLimiter.isQuotaExceeded(
+        userId, blockchainState.getLastBlockHeight(),
+      );
+
+      if (isQuotaExceeded) {
+        const quotaExceededResponse = new ResponseCheckTx();
+
+        quotaExceededResponse.code = AbciError.CODES.RATE_LIMITER_QUOTA_EXCEEDED;
```
Comment (shumkov) [inline lib/abci/handlers/checkTxHandlerFactory.js]: Let's create a new error type `RateLimiterQuotaExceededAbciError` https://github.com/dashevo/js-machine/blob/v0.3-dev/lib/abci/errors/

```
message = `User ${userId} is banned for some time`
data = { userId } 
tags = { key: value, key1: value } // Wrap into KVPair internally
```

File: .env.example [hunk 7]
```diff
@@ -18,3 +18,16 @@ IDENTITY_LEVEL_DB_FILE=./db/identity
 
 # DPP Data Provider cache size for contracts
 DPP_CONTRACT_CACHE_SIZE=500
+
+# Tendermint host and port settings
+# to retrieve rate limiter tags
+TENDERMINT_HOST=127.0.0.1
```
Comment (shumkov) [inline .env.example]: `TENDERMINT_RPC_HOST`

File: lib/abci/errors/RateLimiterQuotaExceededAbciError.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+const AbciError = require('./AbciError');
+
+class RateLimiterQuotaExceededAbciError extends AbciError {
+  /**
+   * @param {Object} data
+   * @param {{ key: string, value: string }[]=[]} tags
+   */
+  constructor(data, tags = []) {
```
Comment (shumkov) [inline lib/abci/errors/RateLimiterQuotaExceededAbciError.js]: We want user mandatory so let's add it as a separate param.

File: lib/abci/errors/RateLimiterUserIsBannedAbciError.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+const AbciError = require('./AbciError');
+
+class RateLimiterUserIsBannedAbciError extends AbciError {
+  /**
+   * @param {Object} data
+   */
+  constructor(data) {
```
Comment (shumkov) [inline lib/abci/errors/RateLimiterUserIsBannedAbciError.js]: The same

File: lib/abci/handlers/checkTxHandlerFactory.js [hunk 10]
```diff
@@ -6,38 +6,74 @@ const {
...
+      }
+
+      const isQuotaExceeded = await rateLimiter.isQuotaExceeded(
+        userId, blockchainState.getLastBlockHeight(),
+      );
+
+      if (isQuotaExceeded) {
+        const banKey = rateLimiter.getBannedKey(blockchainState.getLastBlockHeight());
+
+        throw new RateLimiterQuotaExceededAbciError({ userId }, [
```
Comment (shumkov) [inline lib/abci/handlers/checkTxHandlerFactory.js]: Why not just `{ key: value, key: value }` ?

File: lib/createDIContainer.js [hunk 11]
```diff
@@ -69,6 +82,14 @@ async function createDIContainer(options) {
     stateLevelDBFile: asValue(options.STATE_LEVEL_DB_FILE),
     dppContractCacheSize: asValue(options.DPP_CONTRACT_CACHE_SIZE),
     identityLevelDBFile: asValue(options.IDENTITY_LEVEL_DB_FILE),
+    tendermintHost: asValue(options.TENDERMINT_HOST),
```
Comment (shumkov) [inline lib/createDIContainer.js]: `tendermintRpc...`

File: lib/abci/errors/RateLimiterQuotaExceededAbciError.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+      { userId },
+      tags,
+    );
+
+    this.userId = userId;
+  }
+
+  /**
+   * Get user id
+   */
```
Comment (shuplenkov) [inline lib/abci/errors/RateLimiterQuotaExceededAbciError.js]: `@return {string}`

File: lib/abci/errors/RateLimiterUserIsBannedAbciError.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+      `Identity ${userId} is banned for some time`,
+      { userId },
+    );
+
+    this.userId = userId;
+  }
+
+  /**
+   * Get user id
+   */
```
Comment (shuplenkov) [inline lib/abci/errors/RateLimiterUserIsBannedAbciError.js]: @return {string}

File: lib/createDIContainer.js [hunk 14]
```diff
@@ -69,6 +82,14 @@ async function createDIContainer(options) {
     stateLevelDBFile: asValue(options.STATE_LEVEL_DB_FILE),
     dppContractCacheSize: asValue(options.DPP_CONTRACT_CACHE_SIZE),
     identityLevelDBFile: asValue(options.IDENTITY_LEVEL_DB_FILE),
+    tendermintRpcHost: asValue(options.TENDERMINT_RPC_HOST),
+    tendermintRpcPort: asValue(options.TENDERMINT_RPC_PORT),
+    rateLimiterActive: asValue((options.RATE_LIMITER_ACTIVE === 'true')),
+    rateLimiterMax: asValue(parseInt(options.RATE_LIMITER_MAX_TRANSITIONS_PER_ID, 10)),
```
Comment (shuplenkov) [inline lib/createDIContainer.js]: Maybe it's better to move these numbers to constants or to env variables?

File: lib/abci/handlers/checkTxHandlerFactory.js [hunk 15]
```diff
@@ -6,38 +6,74 @@ const {
...
    *
    * @typedef checkTxHandler
    *
    * @param {abci.RequestCheckTx} request
-   * @return {Promise<abci.ResponseCheckTx>}
+   * @param {BlockchainState} blockchainState
+   * @returns {Promise<abci.ResponseCheckTx>}
    */
-  async function checkTxHandler({ tx: stateTransitionByteArray }) {
+  async function checkTxHandler({ tx: stateTransitionByteArray }, blockchainState) {
```
Comment (shuplenkov) [inline lib/abci/handlers/checkTxHandlerFactory.js]: Move `blockchainState` to factory parameters

File: test/unit/abci/handlers/checkTxHandlerFactory.spec.js [hunk 16]
```diff
@@ -6,38 +6,72 @@ const {
...
+
+  afterEach(async () => {
+    await db.clear();
+    await db.close();
   });
 
-  it('should validate State Transition and return response with code 0', async () => {
-    const response = await checkTxHandler(request);
+  it('should validate a State Transition and return response with code 0', async () => {
+    const response = await checkTxHandler(request, blockchainState);
```
Comment (shuplenkov) [inline test/unit/abci/handlers/checkTxHandlerFactory.spec.js]: Move `blockchainState` to factory, like in my comment above

**Pull request merged**